### DISCUSSION TITLE
[infinity] Scalability of AI-Infinify

### DISCUSSION THREAD

**Discussion by nainamd77** [infiniflow/infinity#444] [Q&A] [answered]
Very much interested in this AI Database AI-Infinity, after knowing the powerful features of it. We need some extra inputs about scalability of this database in production environment, 
1. what are the options this db provides to scale up based on volume of traffic or transactions. 
3. Also is this database cloud based ?. 
Please share information on above doubts.

**Comment by writinwaters** [comment, marked as answer]
Hiya, Thanks for your interest in our project. At the moment, we offer a standalone version only. However, we plan on introducing a distributed version of Infinity in 24Q2 and a cloud-based version in Q3. The distributed version of Infinity scales out based on the volume of traffic rather than transactions.

Cheers,
Writinwaters

**Reply by nainamd77** [reply]
Thanks for info. Also, any executables to install in windows environment? or any plans to do so?

**Reply by writinwaters** [reply]
At the moment we support Linux only. However, we *do* have plans to introduce native support for macOS and Windows in the near future. At the moment, I would recommend utilizing Docker installation on Windows as a workaround.  

Cheers,
Writinwaters

**Reply by nainamd77** [reply]
hi, we have tried docker(docker desktop - windows 10) to use infinity db, pulled and run the image successfully.
but it stopped once started. did not get any error or exception. Any suggestion on this ? or any document about docker installation on windows. need your help.

**Reply by yuzhichang** [reply]
Did you check if another image such as [ubuntu](https://hub.docker.com/_/ubuntu) work at your env?
Screen capture of your procedure would be helpful.
We have not tried Infinity db docker on windows yet, sorry.

**Reply by nainamd77** [reply]
hi, sorry i was away with different tasks. now back to infinity.
Procedure: 1. Pulled infiniflow/infinity:nightly 2. run 
But it exited with code 132.
I tried ubuntu as you suggested, it exited with code 0.
Also, I could view command argument in the files from docker desktop, it is 'null' only where as other images which are running successfully has some valid command parameter like "Cmd": [	"nginx",	"-g",	"daemon off;"	]. please check the attached image for configurations.
![inf_error](https://github.com/infiniflow/infinity/assets/86471730/58063406-7c74-40a6-9623-19c2e3b45890)

**Reply by nainamd77** [reply]
update: ubuntu container now running successfully when started using this command 'docker run -d ubuntu tail -f /dev/null'. but using same command 'docker run -d infiniflow/infinity:nightly tail -f /dev/null' infinity existed with same error 132 code.  please guide us.

**Reply by JinHai-CN** [reply]
@nainamd77 

I also try to start the docker container on my Windows OS. It is running successfully. Screenshot: 
<img width="1211" alt="screenshot 2024-02-04 075136" src="https://github.com/infiniflow/infinity/assets/33142505/5119b5f2-78fb-47e5-b34d-dc4cd5f6da57">
Since infinity need the host CPU support AVX2 instruction set. Would you please check What your CPU types is? I use tool like CPU-Z to get the my laptop info as follow:
	Number of cores		12 (max 12)
	Number of threads	16 (max 16)
	Hybrid			yes, 2 coresets
	Core Set 0		P-Cores, 4 cores, 8 threads
	Core Set 1		E-Cores, 8 cores, 8 threads
	Manufacturer		GenuineIntel
	Name			Intel Core i5 12500H
        ...
        Instructions sets	MMX, SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, EM64T, AES, AVX, **AVX2**, FMA3, SHA

**Reply by nainamd77** [reply]
Thanks for replying, as per your guidance we have found that our system did not have AVX2 (it has AVX only). So we are trying to get another with avx2, I will get back after testing in it.

**Reply by nainamd77** [reply]
hi, finally able to run the container in another system having avx2. But when we trying to connect using python based on sample code given, got errors as below
Could not connect to any of [('127.0.0.1', 23817)] - when using REMOTE_HOST
Could not connect to any of [('0.0.0.0', 23817)] - when using LOCAL_HOST

### Actual error list as below
An error occurred: Could not connect to any of [('0.0.0.0', 23817)]
Exception ignored in: <function RemoteThriftInfinityConnection.__del__ at 0x000001F14DF840E0>  
Traceback (most recent call last):
  File "C:\Python311\Lib\site-packages\infinity\remote_thrift\infinity.py", line 30, in __del__
    if self._is_connected is True:
       ^^^^^^^^^^^^^^^^^^
AttributeError: 'RemoteThriftInfinityConnection' object has no attribute '_is_connected'

both docker and python running in same server.

###   code
import infinity
import infinity.index 
from infinity.common import REMOTE_HOST
from infinity.common import LOCAL_HOST

class dbhandler:
    def __init__(self):
        pass
    def manage_db_req(self):  # Added self parameter here
        try:
            #infinity_obj = infinity.connect(REMOTE_HOST)
            infinity_obj = infinity.connect(LOCAL_HOST)
            db = infinity_obj.get_database("default")
            # Drop my_table if it already exists
            db.drop_table("my_table", if_exists=True)
            # Create a table named "my_table"
            table = db.create_table("my_table", {"num": "integer", "body": "varchar", "vec": "vector, 4, float"}, None)
            table.insert([{"num": 1, "body": "unnecessary and harmful", "vec": [1.0, 1.2, 0.8, 0.9]}])
            table.insert([{"num": 2, "body": "Office for Harmful Blooms", "vec": [4.0, 4.2, 4.3, 4.5]}])
            res = table.output(["*"]).knn("vec", [3.0, 2.8, 2.7, 3.1], "float", "ip", 2).to_pl()
            print(res)
        except Exception as e:
            print(f"An error occurred: {e}")
### docker log
2024-02-07 12:45:09  - profile_record_capacity: 100
2024-02-07 12:45:09  - listen address: 0.0.0.0
2024-02-07 12:45:09  - postgres port: 5432
2024-02-07 12:45:09  - http port: 8088
2024-02-07 12:45:09  - sdk port: 23817
2024-02-07 12:45:09  - log_file_path: /tmp/infinity/log/infinity.log
2024-02-07 12:45:09  - log_to_stdout: false
2024-02-07 12:45:09  - log_max_size: 1.00GB
2024-02-07 12:45:09  - log_file_rotate_count: 10
2024-02-07 12:45:09  - log_level: Info
2024-02-07 12:45:09  - data_dir: /tmp/infinity/data
2024-02-07 12:45:09  - wal_dir: /tmp/infinity/wal
2024-02-07 12:45:09  - default_row_size: 8192
2024-02-07 12:45:09  - buffer_pool_size: 4.00GB
2024-02-07 12:45:09  - temp_dir: /tmp/infinity/temp
2024-02-07 12:45:09  - full_checkpoint_interval_sec: 60
2024-02-07 12:45:09  - full_checkpoint_txn_interval: 0
2024-02-07 12:45:09  - delta_checkpoint_interval_sec: 20
2024-02-07 12:45:09  - delta_checkpoint_interval_wal_bytes: 1000
2024-02-07 12:45:09  - wal_size_threshold: 10240
2024-02-07 12:45:09  - dictionary_dir: /tmp/infinity/resource
2024-02-07 12:45:09 API server listen on: 0.0.0.0:23817, thread pool: 128

please guide us to complete.

**Comment by yuzhichang** [comment]
@nainamd77 Docker Desktop on Windoes doesn't support host network according to https://github.com/docker/roadmap/issues/238. So the infinity-sdk cannot connect to 127.0.0.1:23817 where the container listens.

I managed to run the infinity database and the python sdk **BOTH INSIDE wsl2**:
1. Install wsl2 and Ubuntu 22.04 according to https://learn.microsoft.com/en-us/windows/wsl/install
2. Enable systemd inside wsl2 according to https://learn.microsoft.com/en-us/windows/wsl/systemd
3. Install docker-ce according to https://docs.docker.com/engine/install/ubuntu
4. Start infinify database inside ubuntu:
```
docker pull infiniflow/infinity:nightly
docker run -d --name infinity --network=host infiniflow/infinity:nightly
```
5. Install Python 3.12 inside ubuntu(infinity-sdk requires python 3.10+, however Ubuntu 22.04 has python 3.8):
```
wget https://www.python.org/ftp/python/3.12.2/Python-3.12.2.tar.xz
apt install build-essential libbz2-dev libreadline-dev tk-dev libssl-dev libncurses5-dev libgdbm-dev liblzma-dev sqlite3 libsqlite3-dev zlib1g-dev libffi-dev
tar xJvf Python-3.12.2.tar.xz
cd Python-3.12.2
./configure && make -j 8 && make install
/usr/local/bin/pip3 install infinity-sdk
```
6. Connect to infinity database with infinity-sdk:
```
/usr/local/bin/python3
>>> import infinity
>>> import infinity.index as index
>>> from infinity.common import REMOTE_HOST
>>> infinity_obj = infinity.connect(REMOTE_HOST)
>>> db = infinity_obj.get_database("default")
```

**Reply by nainamd77** [reply]
hi, thanks for guidance. finally we have able to run Infinity in linux - ubuntu system. Now we tested python samples successfully.
 Also we want to test SQL like queries, need guidance to do this. Any tool available with infinity to test this ?.

**Comment by JinHai-CN** [comment]
> hi, thanks for guidance. finally we have able to run Infinity in linux - ubuntu system. Now we tested python samples successfully. Also we want to test SQL like queries, need guidance to do this. Any tool available with infinity to test this ?.

You can find Infinity SQL test cases in https://github.com/infiniflow/infinity/tree/main/test/sql. We use sqllogictest to run these cases. 
Original sqllogictest can be found: https://sqlite.org/sqllogictest/doc/trunk/about.wiki, I would like to use this version: https://github.com/risinglightdb/sqllogictest-rs. Of course, you also can use postgres-client to connect Infinity to run SQL statements.

In fact, we think it is not intuitive to use SQL to do full-text and vector search. Therefore, we do not intend to prioritize support for SQL very high.  And we would like to provide RESTful API and other languages SDK such as Go, Java, Javascript, and C#.

Which kind of API do you want? What features do you want in your scenarios?

**Comment by yuzhichang** [comment]
https://docs.docker.com/network/drivers/host/#docker-desktop
Host networking is also supported on Docker Desktop version 4.29 and later for Mac, Windows, and Linux as a [beta feature](https://docs.docker.com/release-lifecycle/#beta). To enable this feature, navigate to the Features in development tab in Settings, and then select Enable host networking.
